poc/find_wr_params.sage

```diff
@@ -60,252 +81,426 @@ def flp_sound(valid: PineValid):
+    min_num_wr_successes = 0
+    max_num_wr_successes = num_wr_checks + 1
+    num_wr_successes = None
+    while min_num_wr_successes != max_num_wr_successes:
+        mid_num_wr_successes = min_num_wr_successes + \
+            (max_num_wr_successes - min_num_wr_successes) // 2
+        prob = wr_sound(num_wr_checks, mid_num_wr_successes)
+        if bits_of_security(prob) < target_soundness_bits - 1:
+            # Doesn't satisfy soundness, increase the lower bound.
+            min_num_wr_successes = mid_num_wr_successes + 1
+        else:
+            # Satisfies soundness, record this `mid_num_wr_successes` and
+            # decrease the upper bound.
+            num_wr_successes = mid_num_wr_successes
+            max_num_wr_successes = mid_num_wr_successes
+    if num_wr_successes is None:
+        raise ValueError("The number of wraparound checks is too small to "
+                         "satisfy soundness error requirement.")
+    return num_wr_successes
+
+def search_wr_params(target_soundness_bits,
+                     target_zk_bits,
+                     user_params,
+                     max_num_wr_checks,
+                     num_wr_checks_step):
     '''
-    For reach result for which `eta` is set, print soundness error, ZK error,
-    and overhead for the given L2 norm bound and number of fractional bits. To
-    complete the parameters, we run this on the output of `search()`, make an
-    initial guess of `eta`, and manually tune until the desired ZK is reached.
+    Search for the optimal wraparound check parameters for each set of user
+    parameters in `user_params`. The search logic works as follows:
+    - Find all pairs of `num_wr_checks` and `num_wr_successes` that can satisfy
+      soundness error. We do so by finding the `num_wr_successes`, for each
+      `num_wr_checks` from `target_soundness_bits` up to `max_num_wr_checks`,
+      with a step of `num_wr_checks_step`. For each `num_wr_checks`, the search
+      logic is handled in `search_num_wr_successes`.
+    - Then for each set of parameters in `user_params`, go through the pairs of
+      `num_wr_checks` and `num_wr_successes`, and find a viable `alpha`. For
+      each possible combination of `num_wr_checks`, `num_wr_successes`, and
+      `alpha`, compute the overhead in terms of the number of field elements
+      needed to encode the wraparound check results. We pick the combination
+      that gives the lowest overhead.
     '''
+    # The search range for `alpha`.
+    # We pick this maximum alpha, because it ensures a sufficiently low ZK error
+    # for all user parameters.
+    max_alpha = 10
+    min_alpha = compute_min_alpha()
+
+    # Find all pairs of `num_wr_checks` and `num_wr_successes`.
+    num_wr_checks_and_successes = []
+    for num_wr_checks in range(target_soundness_bits,
+                               max_num_wr_checks + 1,
+                               num_wr_checks_step):
```

Comment:
Not immediately obvious to the reader why this `range()` is valid. (Answer: one more wraparound check results in one additional bit of soundness?)